Throw when device does not support USM allocation

lbushi25 opened the pull request:
The USM allocation functions are changed to conform with the spec and throw when certain usm allocation modes are not supported.
malloc_shared is left incomplete because it breaks a lot of tests and its more reasonable to handle that in a later PR. In the meantime, malloc_device and malloc_host have been changed to conform with the spec and throw an exception when the corresponding USM aspect is not supported. The test usm_pooling.cpp has also been updated to reflect this query about the memory access properties in the L0 API so that it knows to expect it in the output.

lbushi25 commented:
This is in relation to this issue : https://github.com/intel/llvm/issues/9513.
The issue has been fixed however there is a test named sycl/test-e2e/USM/usm_pooling.cpp that fails when tested on Intel L0 GPU. This however seems to be due to the test implementation and not an implementation error on my part. In particular, the test is failing because my changes will make the driver code in the test to emit an extra L0 API call which the test does not expect for some reason and the check then fails when the command at line 8 of the test is run.

againull commented on sycl/source/detail/usm/usm_impl.cpp:
```diff
@@ -59,6 +59,13 @@ void *alignedAllocHost(size_t Alignment, size_t Size, const context &Ctxt,
 #endif
+  const auto &devices = Ctxt.get_devices();
+  if (!std::any_of(devices.begin(), devices.end(), [&](const auto &device) {
+        return device.has(sycl::aspect::usm_device_allocations);
```
Should we check aspect::usm_host_allocations here? I think `sycl::aspect::usm_device_allocations` is a mistake.

lbushi25 replied:
Yes, good catch!

againull commented on sycl/source/detail/usm/usm_impl.cpp:
```diff
@@ -59,6 +59,13 @@ void *alignedAllocHost(size_t Alignment, size_t Size, const context &Ctxt,
+        return device.has(sycl::aspect::usm_device_allocations);
+      })) {
+    throw sycl::exception(sycl::errc::feature_not_supported,
+                          "Device does not support Unified Shared Memory!");
```
Can we be more specific and say that usm host allocations are not supported.

againull commented on sycl/source/detail/usm/usm_impl.cpp:
```diff
@@ -131,6 +138,11 @@ void *alignedAllocInternal(size_t Alignment, size_t Size,
                            const context_impl *CtxImpl,
                            const device_impl *DevImpl, alloc Kind,
                            const property_list &PropList) {
+  if ((Kind == alloc::device || Kind == alloc::shared) &&
+      !DevImpl->has(sycl::aspect::usm_device_allocations)) {
+    throw sycl::exception(sycl::errc::feature_not_supported,
+                          "Device does not support Unified Shared Memory!");
+  }
```
It seems that there is the same mistake at this part.
We have three aspects:
usm_host_allocations
usm_device_allocations
usm_shared_allocations

I think we need to check corresponding aspect depending on Kind argument and throw if there is a lack of support.

lbushi25 commented:
@againull 
It seems like my latest change is breaking a number of tests due to throwing an exception when usm_shared_allocations is not supported. The reason seems to be that many of these tests do not catch exceptions when attempting to allocate memory rather they rely on the returned pointer being null. Thoughts on how to move forward?

againull commented:
> @againull I have implemented your suggested corrections. However, it seems like my latest change is breaking a number of tests due to throwing an exception when usm_shared_allocations is not supported. The reason seems to be that many of these tests do not catch exceptions when attempting to allocate memory rather they rely on the returned pointer being null. Thoughts on how to move forward?

If tests are really relying on the behavior that these functions return nullptr when aspect is not supported then we have to fix all these tests in my opinion. Because such behaviour is not aligned with specification.

Per SYCL 2020 spec:
malloc_device:
Throws a synchronous exception with the errc::feature_not_supported error code if the syclDevice does not have aspect::usm_device_allocations. 

malloc_shared:
Throws a synchronous exception with the errc::feature_not_supported error code if the syclDevice does not have aspect::usm_shared_allocations. 

malloc_host:
Throws a synchronous exception with the errc::feature_not_supported error code if the syclDevice does not have aspect::usm_host_allocations. 


So,

againull commented on sycl/source/detail/usm/usm_impl.cpp:
```diff
@@ -252,6 +265,10 @@ void *alignedAlloc(size_t Alignment, size_t Size, const context &Ctxt,
   PrepareNotify.scopedNotify(
       (uint16_t)xpti::trace_point_type_t::mem_alloc_begin);
 #endif
+  if (Kind == alloc::shared && !Dev.has(sycl::aspect::usm_shared_allocations)) {
+    throw sycl::exception(sycl::errc::feature_not_supported,
+                          "Device does not support USM shared allocations!");
+  }
```
Can we move this to alignedAllocInternal? I think it's better to have both checks nearby.

againull commented on sycl/source/detail/usm/usm_impl.cpp:
```diff
@@ -59,6 +59,14 @@ void *alignedAllocHost(size_t Alignment, size_t Size, const context &Ctxt,
+      })) {
+    throw sycl::exception(
+        sycl::errc::feature_not_supported,
+        "No device in this context supports host allocations!");
```
```suggestion
        "No device in this context supports USM host allocations!");
```

lbushi25 commented:
> > @againull I have implemented your suggested corrections. However, it seems like my latest change is breaking a number of tests due to throwing an exception when usm_shared_allocations is not supported. The reason seems to be that many of these tests do not catch exceptions when attempting to allocate memory rather they rely on the returned pointer being null. Thoughts on how to move forward?
> 
> If tests are really relying on the behavior that these functions return nullptr when aspect is not supported then we have to fix all these tests in my opinion. Because such behaviour is not aligned with specification.
> 
> Per SYCL 2020 spec: malloc_device: Throws a synchronous exception with the errc::feature_not_supported error code if the syclDevice does not have aspect::usm_device_allocations.
> 
> malloc_shared: Throws a synchronous exception with the errc::feature_not_supported error code if the syclDevice does not have aspect::usm_shared_allocations.
> 
> malloc_host: Throws a synchronous exception with the errc::feature_not_supported error code if the syclDevice does not have aspect::usm_host_allocations.

@againull Is there someone we can tag to have a closer look at these tests?
For convencience here is the failure link: https://github.com/intel/llvm/actions/runs/7630226166/job/20787344943?pr=12446

lbushi25 commented:
> > > > @againull I have implemented your suggested corrections. However, it seems like my latest change is breaking a number of tests due to throwing an exception when usm_shared_allocations is not supported. The reason seems to be that many of these tests do not catch exceptions when attempting to allocate memory rather they rely on the returned pointer being null. Thoughts on how to move forward?
> > > 
> > > 
> > > If tests are really relying on the behavior that these functions return nullptr when aspect is not supported then we have to fix all these tests in my opinion. Because such behaviour is not aligned with specification.
> > > Per SYCL 2020 spec: malloc_device: Throws a synchronous exception with the errc::feature_not_supported error code if the syclDevice does not have aspect::usm_device_allocations.
> > > malloc_shared: Throws a synchronous exception with the errc::feature_not_supported error code if the syclDevice does not have aspect::usm_shared_allocations.
> > > malloc_host: Throws a synchronous exception with the errc::feature_not_supported error code if the syclDevice does not have aspect::usm_host_allocations.
> > 
> > 
> > @againull Is there someone we can tag to have a closer look at these tests? For convencience here is the failure link: https://github.com/intel/llvm/actions/runs/7630226166/job/20787344943?pr=12446
> 
> I believe tests need to be fixed/updated in scope of this PR. I don't think it makes sense to ask authors of each test to fix it. We don't have a single person who is responsible for maintaining the tests.

@againull 
I removed the code for the case when the allocation is shared because almost all the tests are failing because of it. I think this is the best thing to do because having all the tests changed to conform with the spec could be very risky because I have no knowledge as to what the tests are actually testing and expecting and I will most likely introduce bugs in many of these tests. Instead, I will create another github issue that can be looked separately from this PR.

againull reviewed (approved):
LGTM with condition that tracker for malloc_shared will be created.

lbushi25 commented:
@intel/llvm-gatekeepers Can we get this merged?
The 3 failures are all due to the same test which is unrelated to this PR and is being fixed by https://github.com/intel/llvm/pull/12485

lbushi25 commented:
> LGTM with condition that tracker for malloc_shared will be created.

A github issue has been created here: https://github.com/intel/llvm/issues/12490
Will soon create a JIRA ticket.
